src/databricks/labs/ucx/azure/access.py

```diff
@@ -130,116 +145,210 @@ def _map_storage(self, storage: StorageAccount) -> ValuesView[StoragePermissionM
+                principal_client_id,
+            ),
+            EndpointConfPair(
+                f"spark_conf.fs.azure.account.oauth.provider.type.{storage.name}.dfs.core.windows.net",
+                "org.apache.hadoop.fs.azurebfs.oauth2.ClientCredsTokenProvider",
+            ),
+            EndpointConfPair(
+                f"spark_conf.fs.azure.account.oauth2.client.endpoint.{storage.name}.dfs.core.windows.net",
+                endpoint,
+            ),
+            EndpointConfPair(f"spark_conf.fs.azure.account.auth.type.{storage.name}.dfs.core.windows.net", "OAuth"),
+            EndpointConfPair(
+                f"spark_conf.fs.azure.account.oauth2.client.secret.{storage.name}.dfs.core.windows.net",
+                "{{" + principal_secret_identifier + "}}",
+            ),
+        ]
+
+    def _add_service_principal_configuration_to_workspace_warehouse_config(
+        self,
+        principal_client_id: str,
+        principal_secret_identifier: str,
+        storage_accounts: list[StorageAccount],
+    ):
+        warehouse_config = self._ws.warehouses.get_workspace_warehouse_config()
+        self._installation.save(warehouse_config, filename="warehouse-config-backup.json")
+        sql_dac = warehouse_config.data_access_config or []
+        for storage in storage_accounts:
+            configuration_pairs = self._create_storage_account_data_access_configuration_pairs(
+                principal_client_id, principal_secret_identifier, storage
+            )
+            sql_dac.extend(configuration_pairs)
+        self._ws.warehouses.set_workspace_warehouse_config(
+            data_access_config=sql_dac,
+            sql_configuration_parameters=warehouse_config.sql_configuration_parameters,
+        )
+        logger.info(
+            "Updated workspace warehouse config with service principal connection details for accessing storage accounts"
+        )
+
+    def _remove_service_principal_configuration_from_workspace_warehouse_config(
+        self,
+        principal_client_id: str,
+        principal_secret_identifier: str,
+        storage_accounts: list[StorageAccount],
+    ):
+        try:
+            warehouse_config = self._installation.load(
+                GetWorkspaceWarehouseConfigResponse, filename="warehouse-config-backup.json"
+            )
+        except NotFound:  # For legacy reasons we can not assume the backup to always be present
+            warehouse_config = self._ws.warehouses.get_workspace_warehouse_config()
+        sql_dac = warehouse_config.data_access_config or []
+
+        for storage_account in storage_accounts:
+            configuration_pairs = self._create_storage_account_data_access_configuration_pairs(
+                principal_client_id, principal_secret_identifier, storage_account
             )
+            for configuration_pair in configuration_pairs:
+                if configuration_pair in sql_dac:
+                    sql_dac.remove(configuration_pair)
```

Comment:
does it have a unit test? i'm afraid that `EndpointConfigInfo` may not have `__eq__` and `__hash__` generated for this to work.

Reply:
Good catch! It was not tested, now it is. It works: https://github.com/databrickslabs/ucx/pull/2370/commits/6b6b4d95b30067627ae60fc34391750a792bee47